### PR TITLE
Merge private-key porting series onto const-num-traits experimental base

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary

Merges the two workstreams the fork has been running in parallel:

- **`try_const_num_traits`** — the experimental base that replaced `num-traits` with `const-num-traits` and used git-tagged alphas of `fixed-bigint` / `modmath` / `modmath-cios`.
- **`dev`** — the entire private-key porting series (PRs #18–#38), including the trait-surgery folds, algorithm primitives, encrypt CT gate, and Phase 4/5 legacy removal.

All four upstream crates (`const-num-traits 0.1.1`, `fixed-bigint 0.4.0`, `modmath 0.4.0`, `modmath-cios 0.1.0`) are now released on crates.io, so the `[patch.crates-io]` scaffolding can go.

## Shape

22 commits on top of `try_const_num_traits`:

1. `e89c618` — **Baseline**. Drops both `[patch.crates-io]` blocks (root + `footprint/cortex-m/`), switches `const-num-traits` from the git-tagged alpha to `version = "0.1.1"`, and tightens the `ModMathInt` / `ModMathIntCt` supertrait chain to include `const_num_traits::HasPersonality` (plus `CtIsZero` on the Ct side). Pins the Nct/Ct personality on the impls that route through `modmath::Field<T>` — the released API surface added `<T as HasPersonality>::P = <Nct|Ct>` constraints the alpha didn't need.

2. Twenty-one cherry-picked commits from `dev` (`a991890..dev`) — the private-key porting series in original chronological order:
   - **Algorithm primitives** (PRs #18–#23): `wip-private-key` feature, `rsa_private_op`, `rsa_private_op_and_check`, `pkcs1v15::sign_into`, `pss::sign_into`, defensive-error tests.
   - **Trait-surgery phases** (PRs #24, #25, #26, #33, #36): introduce `GenericPrivateKeyParts`, heapless `GenericSigningKey` wrappers, deprecate and drop the legacy `PrivateKeyParts`.
   - **Storage / wrapper folds** (PRs #27, #28, #29, #31, #32): `Fold-1` through `Fold-2d` — algorithm consumes the generic trait, storage growth, type-alias swaps.
   - **Correctness / gating** (PRs #34, #35, #37, #38): sign_into precision fix, encrypt NCT gate + sealed marker, zeroizing padding switch, comment trim.
   - Skipped: PR #30 (`.pre-commit-config.yaml`) was included as-is.

3. `c3447f0` — Test-side bound fixups for the crates.io API tightening.

## Conflicts encountered during cherry-pick

Two conflicts, both mechanical:

1. `src/modmath_support.rs` at PR #19 — `dev` renamed `pow_loop` → `pow_loop_ct` / `pow_loop_public_exp`; folded that rename into the `HasPersonality<P = Ct>` bound the baseline had.
2. `Cargo.toml` at PR #37 — `dev` bumped `fixed-bigint 0.3.0 → 0.3.2`; kept our `0.4.0`.

## Scope doc

`HEAPLESS_SIGN_ROADMAP.md` at repo root captures the graduation criteria for retiring the `wip-private-key` gate — CT exponent as the load-bearing defense, sequenced Phase 3+ PRs, explicit "no blinding / no on-device CRT construction" scope choice.

## Test plan
- [x] `cargo check` (default: alloc + encoding + modmath)
- [x] `cargo check --no-default-features --features modmath` (heapless)
- [x] `cargo check --no-default-features --features modmath,wip-private-key`
- [x] `cargo test --lib` (97)
- [x] `cargo test --all-features --tests` (17 integration + wycheproof)
- [x] `cargo clippy --all -- -D warnings`

## Summary by Sourcery

Merge the private-key porting work onto the const-num-traits-based backend, introducing generic heapless-compatible RSA key/signing primitives and tightening modular arithmetic traits for CT correctness.

New Features:
- Add a heapless-compatible GenericRsaPrivateKey type and generic PKCS#1 v1.5 / PSS signing key wrappers that work across integer/Montgomery backends.
- Introduce raw rsa_private_op and rsa_private_op_and_check primitives plus heapless PKCS#1 v1.5 and PSS sign_into functions for modular RSA operations.
- Gate CT-capable modulus backends behind a new CtModulusParams marker and use it to restrict encryption APIs to constant-time exponentiation paths.
- Add a wip-private-key feature flag to selectively enable the new generic private-key/signing path on no-alloc builds.

Bug Fixes:
- Fix PSS/PKCS#1 v1.5 signing width checks to use the actual modulus bit-length rather than the integer container precision, avoiding spurious rejections.
- Split CT Montgomery exponentiation into secret- and public-exponent ladders to ensure secret exponents are always processed with constant-time algorithms.
- Tighten ModMathInt / ModMathIntCt bounds with HasPersonality and CtIsZero to match the released const-num-traits API and enforce personality correctness.

Enhancements:
- Refactor RSA private-key handling from the boxed RsaPrivateKey into a backend-generic GenericRsaPrivateKey, with trait-based accessors and zeroize-on-drop semantics.
- Update PSS and PKCS#1 v1.5 encrypting/signing keys to delegate to generic signing/encrypting key types and reuse shared logic.
- Adjust RSA decryption routines to operate over the generic PrivateKeyParts trait, using CRT precompute only when primes and parameters are present.
- Add extensive heapless tests covering rsa_private_op primitives, PKCS#1 v1.5/PSS signing, defensive error paths, and trait satisfaction for generic keys.

Build:
- Switch const-num-traits from a git alpha to the released 0.1.1 crate and drop all [patch.crates-io] overrides for fixed-bigint/modmath/modmath-cios.
- Introduce a pre-commit configuration wiring common linters and Rust tooling (fmt, cargo check, clippy) into the development workflow.

CI:
- Normalize a comment in the minimal-versions CI job while keeping it disabled until pre-release dependencies are removed.

Documentation:
- Add internal documentation and roadmap notes for heapless signing and CT encrypt gating, clarifying scope and security expectations for the new features.

Tests:
- Add heapless RSA test coverage for toy and 2048-bit keys, including round-trip PKCS#1 v1.5/PSS signatures, integrity checks, and regression tests for signing edge cases.
- Extend tests to assert trait bounds, zeroize behavior, and correctness of the new generic signing/private-key abstractions on both Ct and heapless paths.

Chores:
- Tidy license and release configuration files to match the new release process and tooling expectations.

### CODE REVIEW

File: src/algorithms/rsa.rs [hunk 1]
```diff
@@ -54,13 +54,16 @@ where
...
-pub fn rsa_decrypt<R: TryCryptoRng + ?Sized>(
+pub fn rsa_decrypt<R: TryCryptoRng + ?Sized, K>(
     rng: Option<&mut R>,
-    priv_key: &impl PrivateKeyParts<MontyParams = BoxedMontyParams>,
+    priv_key: &K,
     c: &BoxedUint,
-) -> Result<BoxedUint> {
+) -> Result<BoxedUint>
+where
+    K: PrivateKeyParts<BoxedUint, MontyParams = BoxedMontyParams>,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/algorithms/rsa.rs:63]: This bound compiles today (CI on this PR is green, and it's been shipping since Fold-1 in PR #27). `PrivateKeyParts<T>: PublicKeyParts<T>`, and Rust allows projecting a supertrait's associated type in a bound on the subtrait — `PrivateKeyParts<BoxedUint, MontyParams = BoxedMontyParams>` desugars to `where K: PrivateKeyParts<BoxedUint>, <K as PublicKeyParts<BoxedUint>>::MontyParams = BoxedMontyParams`. Not a bug.

File: src/algorithms/rsa.rs [hunk 2]
```diff
@@ -254,6 +266,61 @@ fn unblind(m: &BoxedUint, unblinder: &BoxedUint, n_params: &BoxedMontyParams) ->
...
+/// Raw RSA must be wrapped in a padding/signature scheme (PKCS#1 v1.5, PSS,
+/// OAEP) to be secure. See the [module-level documentation][crate::hazmat]
+/// for more information.
+#[cfg(any(feature = "private-key", feature = "wip-private-key"))]
+#[inline]
+pub fn rsa_private_op<T, M>(c: &T, d: &T, n_params: &M) -> T
+where
+    T: UnsignedModularInt,
+    M: ModulusParams<Modulus = T>,
+    M::MontgomeryForm: Pow<M>,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/algorithms/rsa.rs]: Fixed in 049fa3e. Added `M: CtModulusParams` to `rsa_private_op`, `rsa_private_op_and_check`, `pkcs1v15::sign_into`, `pss::sign_into`, and the wrapping `try_sign_*_into` methods on both `GenericSigningKey<D, T, M>`s. NCT-personality signing keys now fail the trait bound at compile time, same shape as the encrypt gate. The bug predates this merge — it exists on `dev`/`heapless` too — but no need to backport separately since this merge is going to become mainline.
